cli/command_snapshot_fix.go

```diff
@@ -0,0 +1,188 @@
+package cli
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/manifest"
+	"github.com/kopia/kopia/snapshot"
+	"github.com/kopia/kopia/snapshot/snapshotfs"
+)
+
+type commandSnapshotFix struct {
+	removeObjectIDs    []string
+	removeFilesByName  []string
+	fullRead           bool
+	removeInvalidFiles bool
+
+	// snapshots to target
+	manifestIDs []string
+	sources     []string
+
+	commit bool
+
+	verifier *snapshotfs.Verifier
+}
+
+func (c *commandSnapshotFix) setup(svc appServices, parent commandParent) {
+	cmd := parent.Command("fix", "Fix snapshot consistency issues")
+
+	// action to perform
+	cmd.Flag("remove-invalid-files", "Delete invalid files").BoolVar(&c.removeInvalidFiles)
+	cmd.Flag("remove-object", "Remove objects with specified ID").StringsVar(&c.removeObjectIDs)
+	cmd.Flag("remove-file", "Remove file with specified name (unqualified)").StringsVar(&c.removeFilesByName)
+
+	// scope of rewrite
+	cmd.Flag("manifest-id", "Manifest IDs").StringsVar(&c.manifestIDs)
+	cmd.Flag("source", "Source to target (username@hostname:/path)").StringsVar(&c.sources)
+
+	// verification options
+	cmd.Flag("full-read", "Perform full read when verifying files").BoolVar(&c.fullRead)
+
+	cmd.Flag("commit", "Update snapshot manifests").BoolVar(&c.commit)
+
+	cmd.Action(svc.repositoryWriterAction(c.run))
+}
+
+func (c *commandSnapshotFix) rewriteEntries(ctx context.Context, dirRelativePath string, ent *snapshot.DirEntry) (*snapshot.DirEntry, error) {
+	for _, id := range c.removeObjectIDs {
+		if string(ent.ObjectID) == id {
+			return nil, nil
+		}
+	}
+
+	for _, n := range c.removeFilesByName {
+		if ent.Name == n {
+			return nil, nil
+		}
+	}
+
+	if c.verifier != nil {
+		fname := dirRelativePath + "/" + ent.Name
```

Comment:
tiny nit:
```suggestion
		fname := filepath.Join(dirRelativePath, ent.Name)
```

Reply:
FWIW this is actually subtly different (not sure if that matters), on Windows it will use `\` even for Unix snapshots. and `/` and `\` are both supported on Windows, although slash looks a bit out of place.